Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -391,38 +396,89 @@ class Gx2Fitter {
-          // (this allocates storage for all components, we will set them later)
-          auto fittedStates = *result.fittedStates;
-          const auto newTrackIndex = fittedStates.addTrackState(
-              TrackStatePropMask::All, result.lastTrackIndex);
 
-          // now get track state proxy back
-          auto trackStateProxy = fittedStates.getTrackState(newTrackIndex);
-          trackStateProxy.setReferenceSurface(surface->getSharedPtr());
-          // assign the source link to the track state
-          trackStateProxy.setUncalibratedSourceLink(sourcelink_it->second);
+          ACTS_VERBOSE(
+              "Actor - indices before processing:"
+              << "\n\t"
+              << "result.lastMeasurementIndex: " << result.lastMeasurementIndex
+              << "\n\t"
+              << "result.lastTrackIndex: " << result.lastTrackIndex << "\n\t"
+              << "result.fittedStates->size(): " << result.fittedStates->size())
+
+          // TODO generalize the update of the currentTrackIndex
+          size_t currentTrackIndex = Acts::MultiTrajectoryTraits::kInvalid;
+
+          // Checks if an existing surface is found during the gx2f-iteration.
+          // If not, there a new index will be generated afterwards
+          // During the first iteration, we will always create a new index
+          if (nUpdate == 0) {
+            ACTS_VERBOSE("   processSurface: nUpdate == 0 decision");
+
+            // Add a full TrackState entry multi trajectory. This allocates
+            // storage for all components, which we will set later.
+            currentTrackIndex = result.fittedStates->addTrackState(
+                ~(TrackStatePropMask::Smoothed | TrackStatePropMask::Filtered),
+                result.lastTrackIndex);
+          } else {
+            ACTS_VERBOSE("   processSurface: nUpdate > 0 decision");
+
+            if (result.lastTrackIndex ==
+                Acts::MultiTrajectoryTraits::kInvalid) {
+              currentTrackIndex = 0;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (kInv->0) = "
+                           << currentTrackIndex);
+            } else if (result.lastTrackIndex <
+                       result.fittedStates->size() - 1) {
+              currentTrackIndex = result.lastTrackIndex + 1;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (n+1) = "
+                           << currentTrackIndex);
+            } else {
+              // Add a full TrackState entry multi trajectory. This allocates
+              // storage for all components, which we will set later.
+              currentTrackIndex = result.fittedStates->addTrackState(
+                  ~(TrackStatePropMask::Smoothed |
+                    TrackStatePropMask::Filtered),
+                  result.lastTrackIndex);
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (ADD NEW)= "
+                           << currentTrackIndex);
+            }
+          }
```

Comment:
From what I can see the code in the first `if` and the last `else` are the same - could you refactor this to not have the same code twice?

Reply:
I added a new public variable `mask` to the `Actor`. I hope it is now better visible, that it is just a single function call. In the long run, I would like to integrate this also into `getTrackStateProxy()` and set here just the parameters. But first I need to generalise the trackIndex-update.

```
    /// Mask for the track states. We don't need Smoothed and Filtered
    TrackStatePropMask mask =
        ~(TrackStatePropMask::Smoothed | TrackStatePropMask::Filtered);
```